### ISSUE TITLE
Dunst crashes and dumps core when no X server is available

### ISSUE THREAD

**Issue by pmattern** [dunst-project/dunst#1095]
If Dunst is launched in a context where it can't access an X server or Wayland, it crashes and dumps core right away.

This may e. g. happen in a virtual terminal session while X isn't running or when Dunst is launched by root on a system where X is run by a regular user.   
Obviously, neither is an expected use case, but in particular the former can happen by incident rather quickly.   
So all in all I think it would be good, if Dunst handled this situation a bit more gracefully.

Steps to reproduce:
* Make sure that Dunst is the only notification daemon around and a tool to track core dumps is available, e. g. systemd-coredump.
* Have a regular user issue an arbitrary invocation of notify-send on a virtual terminal while no X server or other windowing system is running.
* Check the core dump log. Will be similar to
```
Jul 28 18:58:12 desktop-1 systemd[12638]: Created slice User Core Session Slice.
Jul 28 18:58:12 desktop-1 systemd[12638]: Starting D-Bus User Message Bus...
Jul 28 18:58:12 desktop-1 systemd[12638]: Started D-Bus User Message Bus.
Jul 28 18:58:12 desktop-1 dbus-daemon[12662]: [session uid=1000 pid=12662] Activating systemd to hand-off: service name='org.freedesktop.Notifications' unit='dunst.service' requested by ':1.0' (uid=1000 pid=12660 comm="notify-send --app-name=notify-send CLI Tests zu Du")
Jul 28 18:58:12 desktop-1 dbus-daemon[12662]: [session uid=1000 pid=12662] Successfully activated service 'org.freedesktop.systemd1'
Jul 28 18:58:12 desktop-1 systemd[12638]: Starting Dunst notification daemon...
Jul 28 18:58:12 desktop-1 dunst[12664]: WARNING: Cannot open X11 display.
Jul 28 18:58:12 desktop-1 kernel: show_signal: 6 callbacks suppressed
Jul 28 18:58:12 desktop-1 kernel: traps: dunst[12664] trap int3 ip:7fdc62b26855 sp:7ffe6dbbe590 error:0 in libglib-2.0.so.0.7200.3[7fdc62ae6000+92000]
Jul 28 18:58:12 desktop-1 kernel: audit: type=1701 audit(1659027492.933:174): auid=1000 uid=1000 gid=1000 ses=8 pid=12664 comm="dunst" exe="/usr/bin/dunst" sig=5 res=1
Jul 28 18:58:12 desktop-1 audit[12664]: ANOM_ABEND auid=1000 uid=1000 gid=1000 ses=8 pid=12664 comm="dunst" exe="/usr/bin/dunst" sig=5 res=1
Jul 28 18:58:12 desktop-1 dunst[12664]: ERROR: [  get_x11_output:0065] Couldn't initialize X11 output. Aborting...
Jul 28 18:58:12 desktop-1 systemd[1]: Created slice Slice /system/systemd-coredump.
Jul 28 18:58:12 desktop-1 audit: BPF prog-id=39 op=LOAD
Jul 28 18:58:12 desktop-1 audit: BPF prog-id=40 op=LOAD
Jul 28 18:58:12 desktop-1 audit: BPF prog-id=41 op=LOAD
Jul 28 18:58:12 desktop-1 kernel: audit: type=1334 audit(1659027492.970:175): prog-id=39 op=LOAD
Jul 28 18:58:12 desktop-1 kernel: audit: type=1334 audit(1659027492.970:176): prog-id=40 op=LOAD
Jul 28 18:58:12 desktop-1 kernel: audit: type=1334 audit(1659027492.970:177): prog-id=41 op=LOAD
Jul 28 18:58:12 desktop-1 kernel: audit: type=1130 audit(1659027492.973:178): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-coredump@0-12668-0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
Jul 28 18:58:12 desktop-1 audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-coredump@0-12668-0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
Jul 28 18:58:12 desktop-1 systemd[1]: Started Process Core Dump (PID 12668/UID 0).
Jul 28 18:58:13 desktop-1 systemd[12638]: Starting Monitoring core dumps :: *.service...
Jul 28 18:58:13 desktop-1 systemd-coredump[12669]: Process 12664 (dunst) of user 1000 dumped core.
                                                      
                                                      Module linux-vdso.so.1 with build-id a5290dd75cb1eb0fb20c7a46c6fa011045892af5
                                                      Module libbrotlicommon.so.1 with build-id acfd597a977c8087bb6184383daae2e828a9ce42
                                                      Module libpthread.so.0 with build-id 95ae4f30a6f12ccbff645d30f8e1a3ee23ec7d36
                                                      [...]
```
I took the liberty to quickly post the logs without debug symbols enabled as the problem seems to be 100% and easily reproducible. If you do need an output with debug symbols enabled, just let me know.

1ef38e5 on Arch Linux x86_64 per [AUR](https://aur.archlinux.org/) package [dunst-git](https://aur.archlinux.org/packages/dunst-git).

**Comment by LeCodingWolfie** [comment]
I did also have the same issue, though I am using the release version of [dunst package][Release], and also using Wayland as the main desktop environment through Hyprland as WM. Even more, running dunst-related (`dunstify example` as per the [Wiki introduction][Wiki]) command results in:

> `Unable to send notification: Error calling StartServiceByName for org.freedesktop.Notifications: Timeout was reached`

Trying to restart the daemon shows a similar error and dump. 

Anyways, looking for issues with that headline, an issue of which [referred "duplicate"][Duplicate] [by bebehei][Comment], [can be solved by running:][Solution]

`systemctl --user import-environment DISPLAY`

In fact, I just ran with the issue that a dunst process was running while trying to restart the dunst daemon:

> `CRITICAL: [dbus_cb_name_lost:1044] Cannot acquire 'org.freedesktop.Notifications': Name is acquired by 'dunst' with PID '96228'.` 

which I killed, and could restart the daemon back, and could run `dunstify example`.

Now, it's only showing this warning for me:

> `Xlib:  extension "MIT-SCREEN-SAVER" missing on display ":0".`

Using [this similar solution][Another] may also help to solve the problem.

<sup>By the way, running dunst doesn't output anything, I don't know if it's an actual issue; though it might just start/enable the daemon if it isn't running [(see)][Wiki]</sup> 

[Release]: https://archlinux.org/packages/community/x86_64/dunst/
[Wiki]: https://github.com/dunst-project/dunst/wiki
[Duplicate]: https://github.com/dunst-project/dunst/issues/347
[Comment]: https://github.com/dunst-project/dunst/issues/611#issuecomment-470319797
[Solution]: https://github.com/dunst-project/dunst/issues/347#issuecomment-319370556
[Another]: https://github.com/dunst-project/dunst/issues/611#issuecomment-580393668

**Comment by blocr** [comment]
importing `$WAYLAND_DISPLAY=wayland-0` fixed the crash for me.

**Comment by bynect** [comment]
the core dump is a direct consequence of calling LOG_E(). My question now is, how should we handle this gracefully? Exit with a message? Or wait for a x screen to be set up?

**Comment by nPHYN1T3** [comment]
Not sure what you mean by "wait for an XScreen to be set up" when there are already many. In may case it's not because I'm running Wayland, in my case it's because it's assuming to run on XScreen 0.0 but then called on 0.1 or 0.2. My use case is slowly being killed by the "Wayland future proofing" but Waylands design (at present) breaks by design my hardware config.

However I'd be more likely to vote on exit with a message on start. Waiting for some requirement gives the false illusion things are functioning and the user may continue one assuming things will work.

**Comment by bynect** [comment]
> Not sure what you mean by "wait for an XScreen to be set up" when there are already many. In may case it's not because I'm running Wayland, in my case it's because it's assuming to run on XScreen 0.0 but then called on 0.1 or 0.2. My use case is slowly being killed by the "Wayland future proofing" but Waylands design (at present) breaks by design my hardware config.

sorry, I meant that if the x screen is not found to wait for one

> However I'd be more likely to vote on exit with a message on start. Waiting for some requirement gives the false illusion things are functioning and the user may continue one assuming things will work.

Yes you are totally right and I think this is the right course of action

Maybe the solution could be as easy as changing the LOG_E with DIE (which exits instead of abort)

**Comment by nPHYN1T3** [comment]
I'm guessing from this Dunst will simply not work for me going forward then since it will by design bail rather than just run on the XScreen it's being called from?

**Comment by bynect** [comment]
> I'm guessing from this Dunst will simply not work for me going forward then since it will by design bail rather than just run on the XScreen it's being called from?

If it works now, it will work in the future. I am only talking about using exit instead of abort when reporting the error, so the behavior is not changed

**Comment by nPHYN1T3** [comment]
It doesn't work now which is why I have my above report. Well it works but ONLY on one XScreen.

**Comment by bynect** [comment]
> It doesn't work now which is why I have my above report. Well it works but ONLY on one XScreen. If anything calls it from another XScreen it just dies.

This is the code that opens the display (using xlib) and the only thing where it could fail to initialize like you see.
```c
        if (!(xctx.dpy = XOpenDisplay(NULL))) {
                LOG_W("Cannot open X11 display.");
                return false;
        }
```

this is an excerpt from the xlib documentation
> On a POSIX-conformant system, if the display_name is NULL, it defaults to the value of the DISPLAY environment variable

Thus, dunst is simply trying to open the default display with XOpenDisplay and a NULL. 
This will open the DISPLAY env var. Thus theoretically if you defined DISPLAY correctly this could be solved

**Comment by nPHYN1T3** [comment]
Gotcha, I'm just glad it doesn't crash when called from a different XScreen any longer. Back when I chimed in on this bug report if I launched dunst on say XScreen 0.1 but then an application on XScreen 0.2 did a notification blam, crash. Testing today I see calls actually work.

**Comment by bynect** [comment]
> Gotcha, I'm just glad it doesn't crash when called from a different XScreen any longer. Back when I chimed in on this bug report if I launched dunst on say XScreen 0.1 but then an application on XScreen 0.2 did a notification blam, crash. Testing today I see calls actually work. I had removed a lot of notification stuff from my scripts and tried to silence applications that used it because it just killed the dunst server all the time and only allowed things I knew wouldn't cause the crash.

Well, I'm glad you solved in the end. Who knows, when we refactor the x11 backend we may add support for multiple X servers

**Comment by nPHYN1T3** [comment]
Heh well I know myself a few others would probably dig that but seems nuts like me are a dying breed. Running Multi GPU that is. Or perhaps I should say most multi GPU setups these days are one for output the others for ML / Cuda workloads so they don't care if they are X or Wayland. For those like me who run several screens per GPU Wayland is a death knell unless they fundamentally change how multi GPU works.

**Comment by bynect** [comment]
> Heh well I know myself a few others would probably dig that but seems nuts like me are a dying breed. Maybe XCB would make that easier? (if the documentation doesn't make it harder hah) Running Multi GPU that is. Or perhaps I should say most multi GPU setups these days are one for output the others for ML / Cuda workloads so they don't care if they are X or Wayland. For those like me who run several screens per GPU Wayland is a death knell unless they fundamentally change how multi GPU works.

Yes, using multiple xorg sessions on the same gpu is not something you hear frequently. I am kinda curios, would you mind sharing why/what advantages it has?

Also now that you said xcb, I actually worked with it on another project. so maybe instead of rewriting the xlib backend I can make a xcb backend 🤔

**Comment by nPHYN1T3** [comment]
I'm not running multiple xorg instances (seats) just one xorg instance but with an XScreen per GPU. If you give each GPU it's own XScreen you segregate things. This means I can run say Blender on 3 screens on one GPU (modeling/rendering) while I play a game on another GPU and my game is not bogged down by the blender render, meanwhile I can watch videos on another and there is no cross talk or issues where an xrandr/wayland setup either reduces all GPU's to the weakests' level (and even at idle GPU's push over 50% load for "nothing") or worse Wayland makes all non-primary GPU's worthless slaves to pass frame buffers to making performance even worse since 100% of the work is dumped on whatever GPU is deemed primary.

TL;DR giving each GPU an XScreen ensures all GPU's have 100% capability. Clumping GPU's with a single XScreen/xrandr/Wayland style setup nerfs all GPU performance minimally from 50% to nearly 90%!

**Comment by rickalex21** [comment]
Same here:

-❯ echo $WAYLAND_DISPLAY 
wayland-1

Linux nixos 6.6.26 1-NixOS SMP PREEMPT_DYNAMIC Wed Apr 10 14:36:08 UTC 2024 x86_64 GNU/Linux

-❯ dunst --version
Dunst - A customizable and lightweight notification-daemon 1.10.0

**Comment by bynect** [comment]
> Same here:
> 
> -❯ echo $WAYLAND_DISPLAY wayland-1
> 
> Linux nixos 6.6.26 1-NixOS SMP PREEMPT_DYNAMIC Wed Apr 10 14:36:08 UTC 2024 x86_64 GNU/Linux
> 
> -❯ dunst --version Dunst - A customizable and lightweight notification-daemon 1.10.0

Wait, are you trying to run dunst on wayland and it crashes because there is no X server??

**Comment by nPHYN1T3** [comment]
@rickalex21 bit of a hijack given this is about crashing with a complaint of no X when you are running X. Sounds like you installed from repo's where the repo version is for X. If I'm right you need to check your repos for the correct version for you (Wayland build) or build from source setting the Wayland bit at compile time.

**Comment by rickalex21** [comment]
@bynect Yea I was getting the same errors as mentioned above. It's working fine now. I'm fairly new to
nixos so I'm not sure how dunst is starting considering it's not enabled as a user service in
my configuration.nix, I have no scripts that start it, and nothing in .config/systemd/user.

**Comment by bynect** [comment]
> @bynect Yea I was getting the same errors as mentioned above. It's working fine now. I'm fairly new to
> nixos so I'm not sure how dunst is starting considering it's not enabled as a user service in
> my configuration.nix, I have no scripts that start it, and nothing in .config/systemd/user.

I thought that this was fixed before, I will check again

**Comment by rickalex21** [comment]
So I think I figured out how to reproduce the problem. I cant restart/start the systemd user service
as a user. I am not sure if I have to add xwayland or something to the settings.

```text
Apr 17 09:24:58 nixos dunst[72505]: WARNING: Cannot open X11 display.
Apr 17 09:24:58 nixos dunst[72505]: ERROR: [  get_x11_output:0065] Couldn't initialize X11 output. Aborting...
Apr 17 09:24:58 nixos systemd-coredump[72511]: [🡕] Process 72505 (.dunst-wrapped) of user 1000 dumped core.
```

I used the service file from dunst with some tweaks. Until this gets fixed there's no work around that
I know of other than killing the service and restarting it or rebooting the machine.

```sh
[Unit]
Description=Dunst notification daemon
Documentation=man:dunst(1)
PartOf=default.target

[Service]
Type=dbus
BusName=org.freedesktop.Notifications
ExecStart=/nix/store/f40mb54130x37xfkiqcilz0ylhd1klm4-dunst-1.10.0/bin/dunst

[Install]
WantedBy=default.target

```

I also tried:

```sh
dbus-update-activation-environment DISPLAY WAYLAND_DISPLAY 
```

**Comment by rickalex21** [comment]
@bynect I'm getting `wayland-1` is there anything you can suggest to fix this?

**Comment by bynect** [comment]
> @bynect I'm getting `wayland-1` is there anything you can suggest to fix this?

I am not sure what you mean(?)

Anyhow, dunst checks the WAYLAND_DISPLAY env var and if it is defined tries to use the wayland output (if it was compiled).
[here is the source](https://github.com/dunst-project/dunst/blob/4384521eaaf02ddea68af558dda4546b34a8c4fc/src/output.c#L95C1-L96C1)

It should not even try to use xorg on  a wayland environment (unless force_xwayland is set to true).

**Comment by rickalex21** [comment]
>I am not sure what you mean(?)

I meant the output of `echo $WAYLAND_DISPLAY`

I think it's probably not a big deal since it's a screen saver issue, my main problem was resolved.

```txt
Apr 17 11:36:38 nixos dunst[171679]: Xlib:  extension "MIT-SCREEN-SAVER" missing on display ":0".
```

I'm not familar with the nixos build of dunst but this link may be of interest. This is what comes by default
and what I have: https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/by-name/du/dunst/package.nix

**Comment by bynect** [comment]
> >I am not sure what you mean(?)
> 
> I meant the output of `echo $WAYLAND_DISPLAY` is `wayland-1`.
> 
> I think it's probably not a big deal since it's a screen saver issue, my main problem was resolved.
> 
> ```txt
> Apr 17 11:36:38 nixos dunst[171679]: Xlib:  extension "MIT-SCREEN-SAVER" missing on display ":0".
> ```
> 
> I'm not familar with the nixos build of dunst but this link may be of interest. This is what comes by default
> and what I have: https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/by-name/du/dunst/package.nix
> 

I think that you should not export DISPLAY which is xorg specific when using wayland. I am not too sure what codepath dunst is taking at this point but it should not use Xorg things at all. This message is coming from some library trying to use xorg when there is no xorg.

What happens if you pass only WAYLAND_DISPLAY?

**Comment by rickalex21** [comment]
@bynect I think the solution would be to build dunst without XORG support but not sure I'm up for
that right now since I'm fairly new to nixos.

I don't know what you mean by: `What happens if you pass only WAYLAND_DISPLAY?`

I started getting errors again but I just restart `systemctl --user restart dunst.service` and 
they go away. The only thing that remains is 

```text
Apr 18 14:29:35 nixos systemd[1406]: Starting Dunst notification daemon...
Apr 18 14:29:35 nixos dunst[5176]: Xlib:  extension "MIT-SCREEN-SAVER" missing on display ":0".
Apr 18 14:29:35 nixos systemd[1406]: Started Dunst notification daemon.
```

Which is not a big deal.

**Comment by bynect** [comment]
> @bynect I think the solution would be to build dunst without XORG support but not sure I'm up for that right now since I'm fairly new to nixos.
> 
> I don't know what you mean by: `What happens if you pass only WAYLAND_DISPLAY?`
> 
> I started getting errors again but I just restart `systemctl --user restart dunst.service` and they go away. The only thing that remains is
> 
> ```
> Apr 18 14:29:35 nixos systemd[1406]: Starting Dunst notification daemon...
> Apr 18 14:29:35 nixos dunst[5176]: Xlib:  extension "MIT-SCREEN-SAVER" missing on display ":0".
> Apr 18 14:29:35 nixos systemd[1406]: Started Dunst notification daemon.
> ```
> 
> Which is not a big deal.

Building dunst without Xorg support is a new feature (added in the latest release). The solution should be something else, since dunst should work with both compiled.

Anyway, the DISPLAY env var is for Xorg and WAYLAND_DISPLAY is for wayland. Having both may lead to problems, especially if one is not used. This is why I asked what happens if you don't pass the DISPLAY var at all.

> Apr 18 14:29:35 nixos systemd[1406]: Starting Dunst notification daemon...
Apr 18 14:29:35 nixos dunst[5176]: Xlib:  extension "MIT-SCREEN-SAVER" missing on display ":0".
Apr 18 14:29:35 nixos systemd[1406]: Started Dunst notification daemon.

I suppose that that warning is due to the fact that despite passing the DISPLAY var no Xorg server is running.

But I am not too sure how nixos manages environments.

**Comment by rickalex21** [comment]
@bynect  I took it off from my river `init` but it does not make a difference. That was the only
location where I had DISPLAY at. I only have a single source of truth in ~/.config/fish
and it's not there. I looked at `env`, I think a program I added in nixos is creating it somewhere...

**Comment by rickalex21** [comment]
UPDATE: I did get rid of the errors with "Environment="DISPLAY=:0"

However, now I am getting this:

```sh
Unable to send notification: Error calling StartServiceByName for org.freedesktop.Notifications: Process org.freedesktop.Notifications received signal 5
```

Funny thing about it running "dusnt" directly from the terminal works. More than likely cause I have all my variables in the shell.

**Comment by bynect** [comment]
we no longer coredump after #1466
